packages/driver/src/util/serialization/log.ts

```diff
@@ -0,0 +1,391 @@
+        } as PreprocessedFunction
+      }
+
+      return null
+    }
+
+    return preprocessForSerialization(props)
+  } catch (e) {
+    return null
+  }
+}
+
+/**
+ * Attempts to take in a preprocessed/serialized log-like attributes and reify them. DOM elements are lazily calculated via
+ * getter properties on an object. If these DOM elements are in an array, the array is defined as an ES6 proxy that
+ * ultimately proxies to these getter objects. Functions, if serialized, are rewrapped. If a value cannot be reified for whatever reason,
+ * null is returned.
+ *
+ * This is logLike because there is a need outside of logs, such as in the iframe-model in the runner.
+ * to serialize DOM elements, such as the final snapshot upon request.
+ *
+ * NOTE: this function recursively calls itself to reify a log
+ *
+ * @param {any} props a generic variable that represents a value that has been preprocessed and sent through postMessage() and needs to be reified.
+ * @returns {any} the reified version of the generic.
+ */
+export const reifyLogLikeFromSerialization = (props) => {
+  try {
+    if (props?.serializationKey === 'dom') {
+      props.html = function () {
+        let reifiedElement
+
+        // if snapshots exists, this element needs to be matched against the rendered snapshot on the page LAZILY.
+        // a bit of a hack, but we attach the snapshot to the page and then LAZILY select the element
+
+        const attributes = Object.keys(props.attributes).map((attribute) => {
+          return `[${attribute}="${props.attributes[attribute]}"]`
+        }).join('')
+
+        const selector = `${props.tagName}${attributes}`
+
+        reifiedElement = Cypress.$(selector)
+
+        if (reifiedElement.length) {
+          return reifiedElement.length > 1 ? reifiedElement : reifiedElement[0]
+        }
+
+        // if the element couldn't be found, return a synthetic copy that doesn't actually exist on the page
+        return reifyDomElement(props)
+      }
+
+      return props
+    }
+
+    if (props?.serializationKey === 'function') {
+      const reifiedFunctionData = reifyLogLikeFromSerialization(props.value)
+
+      return () => reifiedFunctionData
+    }
+
+    if (_.isObject(props)) {
+      let reifiedObjectOrArray = {}
+
+      _.forIn(props, (value, key) => {
+        const val = reifyLogLikeFromSerialization(value)
+
+        if (val?.serializationKey === 'dom') {
+          reifiedObjectOrArray = {
+            ...reifiedObjectOrArray,
+            get [key] () {
+              // only calculate the requested $el from console props AFTER the snapshot has been rendered into the AUT
+              return val.html()
+            },
+          }
+        } else {
+          reifiedObjectOrArray[key] = reifyLogLikeFromSerialization(value)
+        }
+      })
+
+      // NOTE: transforms arrays into objects to have defined getters for DOM elements, and proxy back to that object via an ES6 Proxy
```

Comment:
This part is a bit confusing. So objects that have DOM elements are lazily calculated at the time the property is accessed. But this isn't straight forward for arrays. To keep the same lazy calculation, the array is represented as a proxy object. That way array indices "appear" as normal but are really ES6 getters, and the rest of the properties on the array are accessible. A bit strange, but it seems to work really well for our needs.